### ISSUE TITLE
Support of encryption when using Redis from Amazon ElastiCache

### ISSUE THREAD

**Issue by k-rudy** [redis-store/redis-rails#89]
Hey guys,

Amazon ElastiCache allows to set `encryption_in_transit=true` https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/in-transit-encryption.html which enable encryption of the data-on-the-wire. However `redis-rails` is failing to connect to Redis then:

```
Redis::ConnectionError: Connection lost (ECONNRESET)
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:255:in `rescue in io'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:248:in `io'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:259:in `read'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:118:in `block in call'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:229:in `block (2 levels) in process'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:366:in `ensure_connected'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:219:in `block in process'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:304:in `logging'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:218:in `process'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:118:in `call'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:263:in `block in flushdb'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:45:in `block in synchronize'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:45:in `synchronize'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:259:in `flushdb'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-store-1.5.0/lib/redis/store/namespace.rb:75:in `flushdb'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-activesupport-5.0.5/lib/active_support/cache/redis_store.rb:247:in `with'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-activesupport-5.0.5/lib/active_support/cache/redis_store.rb:228:in `block in clear'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/activesupport-5.0.5/lib/active_support/cache.rb:562:in `block in instrument'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/activesupport-5.0.5/lib/active_support/notifications.rb:166:in `instrument'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/activesupport-5.0.5/lib/active_support/cache.rb:562:in `instrument'
+    38s       /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-activesupport-5.0.5/lib/active_support/cache/redis_store.rb:227:in `clear'
``` 

Could you please tell if the gem supports `encryption in transit` feature and if so - what is needed to make it working. Thanks.

**Comment by tubbo** [comment]
@k-rudy Can you verify if the bare `Redis` client connects to your ElastiCache server cluster?

**Comment by k-rudy** [comment]
@tubbo I was able to connect to it via `redis-cli` but through the `stunnel` as documentation suggests - https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/in-transit-encryption.html#connect-tls

Does it mean `stunnel` proxy is also required when using the `redis-rails` with TLS?

**Comment by tubbo** [comment]
@k-rudy I meant the redis.rb client, not the redis-cli client. `Redis::Store` uses the `Redis` Ruby client to connect to Redis, so my hunch is that the problem actually lies within that gem, and not this one. Can you try opening up an `irb` session and typing the following in there?

```ruby
Redis.new('localhost:6379')
```

If you get an error, then I think this issue should be moved to redis/redis-rb, because I'm fairly confident it will need to be fixed within that library.

**Comment by k-rudy** [comment]
@tubbo great point, I get the same error, indeed: 

```ruby
irb(main):006:0> Redis.new(url: 'redis://master....cache.amazonaws.com:6379').get('some_key')
Redis::ConnectionError: Connection lost (ECONNRESET)
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:255:in `rescue in io'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:248:in `io'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:259:in `read'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:118:in `block in call'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:229:in `block (2 levels) in process'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:366:in `ensure_connected'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:219:in `block in process'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:304:in `logging'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:218:in `process'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis/client.rb:118:in `call'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:889:in `block in get'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:45:in `block in synchronize'
	from /usr/lib64/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:45:in `synchronize'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/redis-4.0.1/lib/redis.rb:888:in `get'
	from (irb):6
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/railties-5.0.5/lib/rails/commands/console.rb:65:in `start'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/railties-5.0.5/lib/rails/commands/console_helper.rb:9:in `start'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/railties-5.0.5/lib/rails/commands/commands_tasks.rb:78:in `console'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/railties-5.0.5/lib/rails/commands/commands_tasks.rb:49:in `run_command!'
	from /data/api/shared/bundled_gems/ruby/2.3.0/gems/railties-5.0.5/lib/rails/commands.rb:18:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

moving the issue to `redis.rb`. Appreciate your help

**Comment by tubbo** [comment]
thanks! I follow that project too so I'll watch for your issue :)

**Comment by fonji** [comment]
I just want to add pointers to the solution as my search engine sent me here.
Here's the solved issue https://github.com/redis/redis-rb/issues/771

TL;DR: The solution is to use either `rediss://` (notice the second `s` for `secure`) or add the `ssl: true` option.
